Kernel/Jail.cpp

```diff
@@ -1,21 +1,70 @@
 /*
- * Copyright (c) 2022, Liav A. <liavalb@hotmail.co.il>
+ * Copyright (c) 2022-2023, Liav A. <liavalb@hotmail.co.il>
  *
  * SPDX-License-Identifier: BSD-2-Clause
  */
 
+#include <AK/IntrusiveList.h>
+#include <AK/Singleton.h>
 #include <Kernel/Jail.h>
+#include <Kernel/Process.h>
 
 namespace Kernel {
 
-ErrorOr<NonnullLockRefPtr<Jail>> Jail::create(Badge<JailManagement>, NonnullOwnPtr<KString> name, JailIndex index)
+static Atomic<u64> s_jail_id;
+static Singleton<SpinlockProtected<Jail::List, LockRank::None>> s_all_instances {};
+
+static JailIndex generate_jail_id()
 {
-    return adopt_nonnull_lock_ref_or_enomem(new (nothrow) Jail(move(name), index));
+    return s_jail_id.fetch_add(1);
+}
+
+NonnullRefPtr<ProcessList> Jail::process_list()
+{
+    return m_process_list;
+}
+
+ErrorOr<NonnullLockRefPtr<Jail>> Jail::create(NonnullOwnPtr<KString> name)
+{
+    return s_all_instances->with([&](auto& list) -> ErrorOr<NonnullLockRefPtr<Jail>> {
+        auto process_list = TRY(ProcessList::create());
+        auto jail = TRY(adopt_nonnull_lock_ref_or_enomem(new (nothrow) Jail(move(name), generate_jail_id(), move(process_list))));
+        list.append(jail);
+        return jail;
+    });
+}
+
+ErrorOr<void> Jail::for_each_when_process_is_not_jailed(Function<ErrorOr<void>(Jail const&)> callback)
+{
+    return Process::current().jail().with([&](auto const& my_jail) -> ErrorOr<void> {
+        // Note: If we are in a jail, don't reveal anything about the outside world,
+        // not even the fact that we are in which jail...
+        if (my_jail)
+            return {};
+        return s_all_instances->with([&](auto& list) -> ErrorOr<void> {
+            for (auto& jail : list) {
+                TRY(callback(jail));
+            }
+            return {};
+        });
+    });
+}
+
+LockRefPtr<Jail> Jail::find_by_index(JailIndex index)
```

Comment:
not sure about this; wouldnʼt it be better to provide a callback here to run something immediately when youʼre in the `with` here (to avoid races)?

you also check above whether someone is in the jail but do that separately when using this method… i feel that could be unified for the greater good.

optionally: you could provide a possibility to run over all instances and reuse it in many places with callback that could enable filtering or selecting if possible (you seem to do that few times). of course if that would not incur (too big) performance penalty… 🤔

Reply:
There are use cases for both of those APIs. While I am all for unified checks, we need to consider our use cases and ensure we are not causing headaches to anyone trying to read our code. Therefore for now I will leave this simply like this.